### DISCUSSION TITLE
How to optimally read many parquet files with union_by_name from S3?

### DISCUSSION THREAD

**Discussion by srpwnd** [duckdb/duckdb#10732] [Q&A]
Hi, 

we have hundreds of parquet files in S3 that we need to glob and merge together with `union_by_name` flag set to true.

We do queries like this:

```sql
select *
from read_parquet('s3://<bucket_name>/<path>/**/*.parquet', union_by_name=True)
```

And recently we started receiving `HTTP Get Range` errors. I did some digging around and found out that having many many parquet files in combination with `union_by_name` might be an issue due to the amount of schema scans DuckDB needs to do to properly build the final schema of the unioned table.

Unfortunately, the schema of the files is evolving so we cannot rely on unioning by position. And new files are being created every day.

I've noticed that with 0.10.0 release there is new option `schema` for `read_parquet()` function but it seems it can only set the schema based on `field_id` in the parquet files and all of our parquet files have all `field_id`s set to `NULL`. I was hoping that setting this up would allow us to union the files without DuckDB having to scan for the schema but it seems it is not the case.

Would you suggest any optimization of querying these parquet files from DuckDB?

We've manually merged our existing parquet files for now, but it is only a temporary solution before the number of files grows again.

**Comment by harjeet-glean** [comment]
i faced a similar issue, due to the amount of data we have not even a single complex query is working with duck db even with an instance size of 128 GB and 16 cores.
